### DISCUSSION TITLE
Why no texel format for texture_2d, but f32, i32 or u32

### DISCUSSION THREAD

**Discussion by Alphapage** [gpuweb/gpuweb#4535] [Q&A] [answered]
Hello,

I don't understand why only storage textures need a texel format.
It seems to be easier to work with `texture_2d<rgba8snorm>` instead of `texture_2d<f32>`: the first one is really meaningful (despite I would prefer to call it `rgba8sfloat` instead of `norm` to match with `f32` without confusion).
The second example can be all f32 compatible texel formats.
It could even check if texel format is matching between wgsl and WebGPU parameters.

Thank you in advance for your answers.

**Comment by Kangz** [comment, marked as answer]
Storage textures and textures for sampling can use very different hardware paths depending on the GPU. In all GPUs texture sampling uses a texture unit that knows how to interpret the format of texture to return a `f32` / `u32` / `i32`. So all that's needed in WGSL is the type of the return value.

For storage textures you can imagine that on some hardware there is no special handling of storage textures but instead the shader computes the memory location of the texel and does a memory load/store like it would with a storage buffer. To compute the exact position of the texel in memory, the shader needs to know how large each texel is, so the exact format.

Note that in the future it is very likely that there will be an extension to use storage textures without specifying the exact format, as hardware gained capabilities to do that.
